packages/nextjs/src/config/loaders/dataFetchersLoader.ts

```diff
@@ -0,0 +1,137 @@
+    //
+    // Note: We start with all of the possible wrapper lines in the template and delete the ones we don't need (rather
+    // than starting with none and adding in the ones we do need) because it allows them to live in our souce code as
+    // *code*. If we added them in, they'd have to be strings containing code, and we'd lose all of the benefits of
+    // syntax highlighting, linting, etc.
+    else {
+      // We have to look for declarations and exports separately because when we build the SDK, Rollup turns
+      //     export const XXX = ...
+      // into
+      //     const XXX = ...
+      //     export { XXX }
+      findExports(templateAST, functionName).remove();
+      findDeclarations(templateAST, functionName).remove();
+    }
+  }
+
+  return [userAST.toSource(), templateAST.toSource()];
+}
+
+/**
+ * Wrap `getStaticPaths`, `getStaticProps`, and `getServerSideProps` (if they exist) in the given page code
+ */
+export default function wrapDataFetchersLoader(this: LoaderThis<LoaderOptions>, userCode: string): string {
+  // We know one or the other will be defined, depending on the version of webpack being used
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const { projectDir } = this.getOptions ? this.getOptions() : this.query!;
+
+  // For now this loader only works for ESM code
+  // TODO: Can you even write nextjs pages in CJS?
+  if (!isESM(userCode)) {
+    return userCode;
+  }
+
+  // If none of the functions we want to wrap appears in the page's code, there's nothing to do. (Note: We do this as a
+  // simple substring match (rather than waiting until we've parsed the code) because it's meant to be an
+  // as-fast-as-possible fail-fast. It's possible for user code to pass this check, even if it contains none of the
+  // functions in question, just by virtue of the correct string having been found, be it in a comment, as part of a
+  // longer variable name, etc. That said, when we actually do the code manipulation we'll be working on the code's AST,
+  // meaning we'll be able to differentiate between code we actually want to change and any false positives which might
+  // come up here.)
+  if (Object.keys(DATA_FETCHING_FUNCTIONS).every(functionName => !userCode.includes(functionName))) {
+    return userCode;
+  }
+
+  const templatePath = path.resolve(__dirname, '../templates/dataFetchersLoaderTemplate.js');
+  // make sure the template is included when runing `webpack watch`
+  this.addDependency(templatePath);
+
+  const templateCode = fs.readFileSync(templatePath).toString();
+
+  const [modifiedUserCode, modifiedTemplateCode] = wrapFunctions(
+    userCode,
+    templateCode,
+    path.relative(projectDir, this.resourcePath),
+  );
+
+  // Fill in template placeholders
+  let injectedCode = modifiedTemplateCode;
+  for (const { placeholder, alias } of Object.values(DATA_FETCHING_FUNCTIONS)) {
+    injectedCode = injectedCode.replace(placeholder, alias);
+  }
```

Comment:
This can be a single reduce call? Means that we can reduce bugs with mutable vars.

Reply:
I find `reduce` much harder to parse as a reader. Given that this is build-time code (so no bundle size issues), I'd rather go with the option that's so clear no one has to even think about what it means.